src/napi.h

```diff
@@ -0,0 +1,663 @@
+
+    Buffer();
+    Buffer(napi_env env, napi_value value);
+    size_t Length() const;
+    char* Data() const;
+  };
+
+  /*
+   * The NAPI Error class wraps a JavaScript Error object in a way that enables it
+   * to traverse a C++ stack and be thrown and caught as a C++ exception.
+   *
+   * If a NAPI API call fails without executing any JavaScript code (for example due
+   * to an invalid argument), then the NAPI wrapper automatically converts and throws
+   * the error as a C++ exception of type Napi::Error.
+   *
+   * If a JavaScript function called by C++ code via NAPI throws a JavaScript exception,
+   * then the NAPI wrapper automatically converts and throws it as a C++ exception of type
+   * Napi::Error.
+   *
+   * If a C++ exception of type Napi::Error escapes from a NAPI C++ callback, then
+   * the NAPI wrapper automatically converts and throws it as a JavaScript exception.
+   *
+   * Catching a C++ exception of type Napi::Error also clears the JavaScript exception.
+   * Of course it may be then re-thrown, which restores the JavaScript exception.
+   *
+   * Example 1 - Throwing an exception:
+   *
+   *   Napi::Env env = ...
+   *   throw env.NewError("Example exception");
+   *   // Following C++ statements will not be executed.
+   *   // The exception will bubble up as a C++ exception of type Napi::Error,
+   *   // until it is either caught while still in C++, or else automatically
+   *   // re-thrown as a JavaScript exception when the callback returns to JavaScript.
+   *
+   * Example 2 - Ignoring a NAPI exception:
+   *
+   *   Napi::Function jsFunctionThatThrows = someObj.AsFunction();
+   *   jsFunctionThatThrows({ arg1, arg2 });
+   *   // Following C++ statements will not be executed.
+   *   // The exception will bubble up as a C++ exception of type Napi::Error,
+   *   // until it is either caught while still in C++, or else automatically
+   *   // re-thrown as a JavaScript exception when the callback returns to JavaScript.
+   *
+   * Example 3 - Handling a NAPI exception:
+   *
+   *   Napi::Function jsFunctionThatThrows = someObj.AsFunction();
+   *   try {
+   *      jsFunctionThatThrows({ arg1, arg2 });
+   *   }
+   *   catch (const Napi::Error& e) {
+   *     cerr << "Caught JavaScript exception: " + e.what();
+   *     // Since the exception was caught here, it will not be re-thrown as
+   *     // a JavaScript exception.
+   *   }
+   */
+  class Error : public Value, public std::exception {
+  public:
+    static Error New(Napi::Env env);
+    static Error New(Napi::Env env, const char* message);
+    static Error NewTypeError(Napi::Env env, const char* message);
+    static Error NewRangeError(Napi::Env env, const char* message);
+
+    Error();
+    Error(napi_env env, napi_value value);
+
+    std::string Message() const;
+    void ThrowAsJavaScriptException() const;
+
+    const char* what() const override;
+
+  private:
+    std::string _message;
+  };
+
+  class Persistent {
```

Comment:
I am updating this to `Persistent<T>`, where `T` may be any `Value` subclass. The only change will be that the `Value()` method returns a `T` instead of a `Value`. Then each of the `Value` subclasses will have an override of `MakePersistent()` that returns a specific persistent type.